File: slidingwindow/003-longest-substring-without-repeating/LongestSubstringWithoutRepeating.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map <Character,Integer> map  = new HashMap<>();
        int start =0;
        int maxLen =0;
        for(int end =0;end<s.length();end++){
            char current = s.charAt(end);
            if(map.containsKey(current)){
                start = Math.max(start, map.get(current)+1);
            }
            map.put(current, end);
            maxLen = Math.max(maxLen, end-start+1);
        }
        return maxLen;
    }
}

File: slidingwindow/003-longest-substring-without-repeating/LongestSubstringWithoutRepeating.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_index = {}
        start =0
        maxLen =0
        for end in range(len(s)):
            if s[end] in char_index and char_index[s[end]]>=start:
                start = char_index[s[end]]+1
            char_index[s[end]]=end
            maxLen = max(maxLen, end-start+1)
        return maxLen